File: examples/readme/smart-simulation/smart_simulation.mjs
import { writeFileSync } from "node:fs";
import { Schematic, TickSettleMode, TickSimulation } from "nucleation";

// --8<-- [start:author]
const scene = Schematic.create("smart_circuit");
scene.fillCuboid(0, 0, 0, 8, 0, 2, "minecraft:smooth_stone");
scene.setBlock(0, 1, 0, "minecraft:lever[face=floor,facing=east,powered=false]");

// One engine setup, six placements. Each wire sees the state left by the last.
const wirePositions = Array.from({ length: 6 }, (_, i) => [i + 1, 1, 0]).flat();
if (scene.setBlocksSimulated(wirePositions, "minecraft:redstone_wire") !== 6) {
  throw new Error("wire batch changed");
}

scene.setBlock(7, 1, 0, "minecraft:redstone_lamp[lit=false]{simulate=true}");
scene.setBlock(0, 1, 2, "minecraft:barrel[facing=west]{signal=13,item=iron_ingot}");
// --8<-- [end:author]


// --8<-- [start:tick]
const tick = TickSimulation.fromSchematic(scene, TickSettleMode.InWorld, 0, 0, 0, "");
tick.useBlock(0, 1, 0);
tick.run(2);
if (tick.getBlock(7, 1, 0) !== "minecraft:redstone_lamp[lit=true]") {
  throw new Error("tick engine did not light lamp");
}
if (tick.tickCount() !== 2) throw new Error("tick count changed");
// --8<-- [end:tick]


if (scene.getBlockString(3, 1, 0) !==
    "minecraft:redstone_wire[east=side,north=none,power=0,south=none,west=side]") {
  throw new Error("wire state changed");
}
const barrel = JSON.parse(scene.getBlockEntityJson(0, 1, 2));
if (barrel.nbt.Items.List[0].Compound.id.String !== "minecraft:iron_ingot") {
  throw new Error("barrel shorthand changed");
}
if (scene.blockCount() !== 36) throw new Error("block count changed");

const bytes = Uint8Array.from(Buffer.from(scene.toSchematicB64(), "base64"));
const output = process.env.SMART_SIMULATION_OUT ?? "smart-circuit.schem";
writeFileSync(output, bytes);
console.log(`Smart simulation JavaScript example: OK (${output})`);

File: examples/readme/smart-simulation/smart_simulation.py
"""Executable Python source for docs/features/smart-placement-and-simulation.md."""

import json
import os
from pathlib import Path

from nucleation import MchprsWorld, Schematic, TickSettleMode, TickSimulation


# --8<-- [start:author]
scene = Schematic.create("smart_circuit")
scene.fill_cuboid(0, 0, 0, 8, 0, 2, "minecraft:smooth_stone")
scene.set_block(0, 1, 0, "minecraft:lever[face=floor,facing=east,powered=false]")

# One engine setup, six placements. Each wire sees the state left by the last.
wire_positions = [coordinate for x in range(1, 7) for coordinate in (x, 1, 0)]
assert scene.set_blocks_simulated(wire_positions, "minecraft:redstone_wire") == 6

scene.set_block(7, 1, 0, "minecraft:redstone_lamp[lit=false]{simulate=true}")
scene.set_block(0, 1, 2, "minecraft:barrel[facing=west]{signal=13,item=iron_ingot}")
# --8<-- [end:author]


# --8<-- [start:mchprs]
world = MchprsWorld.create(scene)
assert not world.is_lit(7, 1, 0)
world.on_use_block(0, 1, 0)
world.tick(2)
world.flush()
assert world.is_lit(7, 1, 0)
assert world.get_redstone_power(6, 1, 0) == 10
# --8<-- [end:mchprs]


# --8<-- [start:tick]
tick = TickSimulation.from_schematic(scene, TickSettleMode.InWorld, 0, 0, 0, "")
tick.use_block(0, 1, 0)
tick.run(2)
assert tick.get_block(7, 1, 0) == "minecraft:redstone_lamp[lit=true]"
assert tick.tick_count() == 2
# --8<-- [end:tick]


wire = scene.get_block_string(3, 1, 0)
barrel = json.loads(scene.get_block_entity_json(0, 1, 2))
assert wire == "minecraft:redstone_wire[east=side,north=none,power=0,south=none,west=side]"
items = barrel["nbt"]["Items"]["List"]
assert items[0]["Compound"]["id"]["String"] == "minecraft:iron_ingot"
assert scene.block_count() == 36
size = scene.tight_dimensions()
assert (size.x, size.y, size.z) == (9, 2, 3)

output = Path(os.environ.get("SMART_SIMULATION_OUT", "smart-circuit.schem"))
output.parent.mkdir(parents=True, exist_ok=True)
scene.save_to_file(str(output))
print(f"Smart simulation Python example: OK ({output})")
